server/src/main/java/org/elasticsearch/tasks/TaskResultsService.java

```diff
@@ -159,7 +178,13 @@ public void onResponse(IndexResponse indexResponse) {
 
             @Override
             public void onFailure(Exception e) {
-                listener.onFailure(e);
+                if (backoff.hasNext()) {
```

Comment:
Should we be a bit more selective about when we retry here? Like if the failure is `TOO_MANY_REQUESTS` we should definitely retry, but if it is a mapping failure, it's unlikely somebody will fix it within a minute, so we definitely shouldn't. And, to be honest, I cannot think about any other failure besides EsRejectedExecutionException where it would definitely make sense to retry. If the index is temporary unavailable and we are here, that means that we already retired, right? If settings or mappings are screwed up or shard is freaking out for any other reasons, we probably should just give up as well. What do you think?

Reply:
>  If the index is temporary unavailable and we are here, that means that we already retired, right?

Like we try to write to the index and the shard isn't allocated. Like if both of the nodes that had a copy of it went down. That seems fairly unlikely, but possible. I wonder if I should just do rejected execution exception for now and look at that case in a follow up.

Reply:
>  I wonder if I should just do rejected execution exception for now and look at that case in a follow up.

I think this would be a great first step.

Reply:
Have another look now!